### PR TITLE
Feature/add create rubric endpoint

### PR DESCRIPTION
**Pull request by AnnPostnikova**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/1037

### CODE REVIEW

File: src/WikiService.Business/Commands/Rubric/CreateRubricCommand.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      ICreateRubricRequestValidator validator,
+      IHttpContextAccessor httpContextAccessor,
+      IResponseCreator responseCreator)
+    {
+      _repository = repository;
+      _mapper = mapper;
+      _validator = validator;
+      _httpContextAccessor = httpContextAccessor;
+      _responseCreator = responseCreator;
+    }
```
Comment (MariyaSin) [inline src/WikiService.Business/Commands/Rubric/CreateRubricCommand.cs:42]: add empty line

File: src/WikiService.Business/Commands/Rubric/CreateRubricCommand.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    public async Task<OperationResultResponse<Guid?>> ExecuteAsync(CreateRubricRequest request)
+    {
```
Comment (MariyaSin) [inline src/WikiService.Business/Commands/Rubric/CreateRubricCommand.cs:45]: add check right

File: src/WikiService.Data/Interfaces/IRubricRepository.cs [hunk 3]
```diff
@@ -1,9 +1,19 @@
...
+using System.Threading.Tasks;
 
 namespace LT.DigitalOffice.WikiService.Data.Interfaces
 {
   [AutoInject]
   public interface IRubricRepository
   {
+    Task<Guid?> CreateAsync(DbRubric dbRubric);
+
+    Task<bool> DoesExistAsync(Guid rubricParentId);
```
Comment (MariyaSin) [inline src/WikiService.Data/Interfaces/IRubricRepository.cs]: the metod check if rubric by id exist, rename rubricParentId to rubricId or rename method

File: src/WikiService.Data/Interfaces/IRubricRepository.cs [hunk 4]
```diff
@@ -1,9 +1,19 @@
...
+
+    Task<bool> DoesSubrubricNameExistAsync(Guid rubricParentId, string nameRubric);
```
Comment (MariyaSin) [inline src/WikiService.Data/Interfaces/IRubricRepository.cs]: join the method with next

File: src/WikiService.Mappers/Db/DbRubricMapper.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    public DbRubricMapper(
+      IHttpContextAccessor httpContextAccessor)
+    {
+      _httpContextAccessor = httpContextAccessor;
+    }
+
+    public DbRubric Map(CreateRubricRequest request)
+    {
+      if (request == null)
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Db/DbRubricMapper.cs]: please use the "is" operator for a single code style in the project

File: src/WikiService.Mappers/Db/DbRubricMapper.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      {
+        return null;
+      }
+
+      Guid rubricId = Guid.NewGuid();
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Db/DbRubricMapper.cs]: do you need create the variable here?

File: src/WikiService.Mappers/Db/Interfaces/IDbRubricMapper.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿using LT.DigitalOffice.Kernel.Attributes;
+using LT.DigitalOffice.WikiService.Models.Db;
+using LT.DigitalOffice.WikiService.Models.Dto.Requests;
+
+namespace LT.DigitalOffice.WikiService.Mappers.Db.Interfaces
+{
+  [AutoInject]
+  public  interface IDbRubricMapper
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Db/Interfaces/IDbRubricMapper.cs]: fix spases between "public" and  "interface"

File: src/WikiService.Mappers/Db/Interfaces/IDbRubricMapper.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+    {
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Db/Interfaces/IDbRubricMapper.cs]: fix tab

File: src/WikiService.Models.Dto/Requests/CreateRubricRequest.cs [hunk 9]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+
+namespace LT.DigitalOffice.WikiService.Models.Dto.Requests
+{
+  public record CreateRubricRequest
+  {
+    public string Name { get; set; }
```
Comment (MariyaSin) [inline src/WikiService.Models.Dto/Requests/CreateRubricRequest.cs:9]: add requred attribure

File: src/WikiService.Validation/Rubric/CreateRubricRequestValidator.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+{
+  public class CreateRubricRequestValidator : AbstractValidator<CreateRubricRequest>, ICreateRubricRequestValidator
+  {
+    public CreateRubricRequestValidator(
+      IRubricRepository rubricRepository)
+    {
+      CascadeMode = CascadeMode.Stop;
+
+      RuleFor(rubric => rubric.Name.Trim())
+        .NotEmpty()
```
Comment (MariyaSin) [inline src/WikiService.Validation/Rubric/CreateRubricRequestValidator.cs]: use requred attribute do not check empty syting

File: src/WikiService.Validation/Rubric/Interfaces/ICreateRubricRequestValidator.cs [hunk 11]
```diff
@@ -0,0 +1,12 @@
+﻿using FluentValidation;
+using LT.DigitalOffice.Kernel.Attributes;
+using LT.DigitalOffice.WikiService.Models.Dto.Requests;
+
+namespace LT.DigitalOffice.WikiService.Validation.Rubric.Interfaces
+{
+  [AutoInject]
+  public interface ICreateRubricRequestValidator : IValidator<CreateRubricRequest>
+  {
+
```
Comment (MariyaSin) [inline src/WikiService.Validation/Rubric/Interfaces/ICreateRubricRequestValidator.cs]: remove empty line

File: src/WikiService/appsettings.json [hunk 12]
```diff
@@ -42,7 +42,11 @@
 
     // FileService
     "CreateFilesEndpoint": "FileService_CreateFiles",
-  },
+
+    // AuthService
+    "ValidateTokenEndpoint": "AuthService_ValidateToken"
+
```
Comment (MariyaSin) [inline src/WikiService/appsettings.json]: remove empty line

Comment (AnnPostnikova) [reply src/WikiService/appsettings.json]: done

File: src/WikiService.Validation/Rubric/Interfaces/ICreateRubricRequestValidator.cs (hunk 11 above)
Comment (AnnPostnikova) [reply src/WikiService.Validation/Rubric/Interfaces/ICreateRubricRequestValidator.cs]: done

File: src/WikiService.Mappers/Db/Interfaces/IDbRubricMapper.cs (hunk 8 above)
Comment (AnnPostnikova) [reply src/WikiService.Mappers/Db/Interfaces/IDbRubricMapper.cs]: done

File: src/WikiService.Mappers/Db/Interfaces/IDbRubricMapper.cs (hunk 7 above)
Comment (AnnPostnikova) [reply src/WikiService.Mappers/Db/Interfaces/IDbRubricMapper.cs]: done

File: src/WikiService.Mappers/Db/DbRubricMapper.cs (hunk 6 above)
Comment (AnnPostnikova) [reply src/WikiService.Mappers/Db/DbRubricMapper.cs]: changed create the variable to new DbRubric in Id prop

File: src/WikiService.Business/Commands/Rubric/CreateRubricCommand.cs [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+using Microsoft.AspNetCore.Http;
+using System;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace LT.DigitalOffice.WikiService.Business.Commands.Rubric
+{
+  public class CreateRubricCommand : ICreateRubricCommand
+    {
```
Comment (VLiamin) [inline src/WikiService.Business/Commands/Rubric/CreateRubricCommand.cs]: Check tab

**Review by VLiamin — APPROVED** [review]

**Review by AnnaGakh — APPROVED** [review]

**Review by a5x55555 — APPROVED** [review]

File: src/WikiService.Data/Interfaces/IRubricRepository.cs [hunk 14]
```diff
@@ -1,9 +1,17 @@
...
+using System.Threading.Tasks;
 
 namespace LT.DigitalOffice.WikiService.Data.Interfaces
 {
   [AutoInject]
   public interface IRubricRepository
   {
+    Task<Guid?> CreateAsync(DbRubric dbRubric);
+
+    Task<bool> DoesRubricIdExistAsync(Guid rubricId);
```
Comment (ndoriomedov) [inline src/WikiService.Data/Interfaces/IRubricRepository.cs]: you can name it as DoesExistAsync

File: src/WikiService.Validation/Rubric/CreateRubricRequestValidator.cs [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+        .MaximumLength(150)
+        .WithMessage("Rubric name is too long.");
+
+      When(rubric => rubric.ParentId.HasValue, () =>
+      {
+        RuleFor(request => request.ParentId)
+          .NotEmpty()
+          .WithMessage("ParentId must not be empty.")
+          .MustAsync(async (parentId, _) => await rubricRepository.DoesRubricIdExistAsync(parentId.Value))
+          .WithMessage("This rubric id does not exist.")
```
Comment (ndoriomedov) [inline src/WikiService.Validation/Rubric/CreateRubricRequestValidator.cs]: this subrubric doesn't exist

**Review by ndoriomedov — APPROVED** [review]

File: src/WikiService.Validation/Rubric/CreateRubricRequestValidator.cs [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+namespace LT.DigitalOffice.WikiService.Validation.Rubric
+{
+  public class CreateRubricRequestValidator : AbstractValidator<CreateRubricRequest>, ICreateRubricRequestValidator
+  {
+    public CreateRubricRequestValidator(
+      IRubricRepository rubricRepository)
+    {
+      CascadeMode = CascadeMode.Stop;
+
+      RuleFor(rubric => rubric.Name.Trim())
```
Comment (MariyaSin) [inline src/WikiService.Validation/Rubric/CreateRubricRequestValidator.cs]: remove trim

File: src/WikiService.Validation/Rubric/CreateRubricRequestValidator.cs [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+      {
+        RuleFor(request => request.ParentId)
+          .NotEmpty()
+          .WithMessage("ParentId must not be empty.")
+          .MustAsync(async (parentId, _) => await rubricRepository.DoesExistAsync(parentId.Value))
+          .WithMessage("This rubric id doesn't exist.")
+          .DependentRules(() =>
+          {
+            RuleFor(request => request)
+           .MustAsync(async (request, _) => !await rubricRepository.DoesRubricNameExistAsync(request.ParentId.Value, request.Name))
```
Comment (MariyaSin) [inline src/WikiService.Validation/Rubric/CreateRubricRequestValidator.cs]: add tab

File: src/WikiService.Validation/Rubric/CreateRubricRequestValidator.cs [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+           .WithMessage("This subrubric name already exists.");
```
Comment (MariyaSin) [inline src/WikiService.Validation/Rubric/CreateRubricRequestValidator.cs]: add tab

File: src/WikiService.Validation/Rubric/CreateRubricRequestValidator.cs [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+        .WithMessage("Rubric name is too long.");
+
+      When(rubric => rubric.ParentId.HasValue, () =>
+      {
+        RuleFor(request => request.ParentId)
+          .NotEmpty()
+          .WithMessage("ParentId must not be empty.")
+          .MustAsync(async (parentId, _) => await rubricRepository.DoesExistAsync(parentId.Value))
+          .WithMessage("This rubric id doesn't exist.")
+          .DependentRules(() =>
```
Comment (MariyaSin) [inline src/WikiService.Validation/Rubric/CreateRubricRequestValidator.cs]: use  ruleFor in 36 line to check name

**Review by MariyaSin — APPROVED** [review]

**Pull request merged**